From 95396e23cc3d1af40c669a59392640fcdf2fde7c Mon Sep 17 00:00:00 2001
From: Sergio Martins <sergio.martins@kdab.com>
Date: Thu, 12 Jul 2018 18:37:25 +0100
Subject: [PATCH 17/32] Make sure unsetting a stylesheet will restore changed
 properties

With the 'qproperty-' syntax we can change widget's properties through
stylesheet. When clearing the stylesheet those properties are now restored.

Change-Id: Ibf00fd3ff6eef4b3100c1100e7b9bbac25445d61
Reviewed-on: https://codereview.kdab.com/69431
Reviewed-by: Aimini <daimini@amazon.com>
---
 src/widgets/styles/qstylesheetstyle.cpp | 10 ++++++++++
 src/widgets/styles/qstylesheetstyle_p.h | 14 ++++++++++++++
 2 files changed, 24 insertions(+)

diff --git a/src/widgets/styles/qstylesheetstyle.cpp b/src/widgets/styles/qstylesheetstyle.cpp
index 18822f3c32..3bfbdfaf8d 100644
--- a/src/widgets/styles/qstylesheetstyle.cpp
+++ b/src/widgets/styles/qstylesheetstyle.cpp
@@ -2627,6 +2627,8 @@ void QStyleSheetStyle::setProperties(QWidget *w)
         }
 
         w->setProperty(propertyL1, v);
+
+        styleSheetCaches->propertyRollbacks[w].push_back( { w, property.toLatin1(), v });
     }
 }
 
@@ -2801,6 +2803,7 @@ void QStyleSheetStyleCaches::objectDestroyed(QObject *o)
     customFontWidgets.remove(static_cast<QWidget *>(o));
     styleSheetCache.remove(o);
     autoFillDisabledWidgets.remove((const QWidget *)o);
+    propertyRollbacks.remove((const QWidget *)o);
 }
 
 void QStyleSheetStyleCaches::styleDestroyed(QObject *o)
@@ -2963,6 +2966,13 @@ void QStyleSheetStyle::unpolish(QWidget *w)
     styleSheetCaches->renderRulesCache.remove(w);
     styleSheetCaches->styleSheetCache.remove(w);
     unsetPalette(w);
+
+    const auto &propertyRollbacks = styleSheetCaches->propertyRollbacks.value(w);
+    for (PropertyRollback rollback : propertyRollbacks)
+        rollback.rollback();
+
+    styleSheetCaches->propertyRollbacks.remove(w);
+
     setGeometry(w);
     w->setAttribute(Qt::WA_StyleSheetTarget, false);
     w->setAttribute(Qt::WA_StyleSheet, false);
diff --git a/src/widgets/styles/qstylesheetstyle_p.h b/src/widgets/styles/qstylesheetstyle_p.h
index 0a30b2aaec..024492a028 100644
--- a/src/widgets/styles/qstylesheetstyle_p.h
+++ b/src/widgets/styles/qstylesheetstyle_p.h
@@ -179,6 +179,19 @@ private:
     Q_DECLARE_PRIVATE(QStyleSheetStyle)
 };
 
+struct PropertyRollback
+{
+    typedef QVector<PropertyRollback> List;
+    void rollback()
+    {
+        widget->setProperty(propertyName, oldValue);
+    }
+
+    QWidget *widget;
+    QByteArray propertyName;
+    QVariant oldValue;
+};
+
 class QStyleSheetStyleCaches : public QObject
 {
     Q_OBJECT
@@ -213,6 +226,7 @@ public:
     };
     QHash<const QWidget *, Tampered<QPalette>> customPaletteWidgets;
     QHash<const QWidget *, Tampered<QFont>> customFontWidgets;
+    QHash<const QWidget *, PropertyRollback::List> propertyRollbacks;
 };
 template <typename T>
 class QTypeInfo<QStyleSheetStyleCaches::Tampered<T>>
-- 
2.28.0.windows.1

